### PR TITLE
Let core.get_mod_storage be called multiple times

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
This is a bug fix. If a single mod calls `core.get_mod_storage` multiple times, it should get the same reference every time, unless references are garbage-collected between calls. Without this PR, `core.get_mod_storage` cannot be called multiple times unless references are garbage-collected between calls.

This PR works by implementing `core.get_mod_storage` in Lua. The C++ version still exists, but it takes the desired mod name as an argument, and is only available internally.

## To do

This PR is Ready for Review.

## How to test

Have this code run at some mod's initialization time:

```lua
assert(core.get_mod_storage() == core.get_mod_storage())                        
collectgarbage()                                                                
assert(core.get_mod_storage())
```

### CODE REVIEW

File: builtin/common/storage.lua [hunk 1]
```diff
@@ -0,0 +1,19 @@
+-- Modify core.get_mod_storage to return the storage for the current mod.
+
+local get_current_modname = core.get_current_modname
+
+local old_get_mod_storage = core.get_mod_storage
+
+local storages = setmetatable({}, {
+	__mode = "v", -- Storages can be remade after being garbage-collected.
```
Comment (sfan5) [inline builtin/common/storage.lua:8]:
```suggestion
	__mode = "v", -- values are weak references (can be garbage-collected)
```

File: builtin/game/init.lua [hunk 2]
```diff
@@ -17,6 +17,7 @@ if core.settings:get_bool("profiler.load") then
 end
 
 dofile(commonpath .. "after.lua")
+dofile(commonpath .. "storage.lua")
```
Comment (sfan5) [inline builtin/game/init.lua:20]: maybe name it modstorage.lua?

File: src/script/lua_api/l_storage.cpp [hunk 3]
```diff
@@ -25,12 +25,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 int ModApiStorage::l_get_mod_storage(lua_State *L)
 {
-	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_CURRENT_MOD_NAME);
-	if (!lua_isstring(L, -1)) {
-		return 0;
-	}
-
-	std::string mod_name = readParam<std::string>(L, -1);
+	std::string mod_name = readParam<std::string>(L, 1);
```
Comment (sfan5) [inline src/script/lua_api/l_storage.cpp:28]:
```suggestion
	// Note that this is wrapped in Lua, see builtin/[...].lua
	std::string mod_name = readParam<std::string>(L, 1);
```

File: src/script/lua_api/l_storage.cpp [hunk 4]
```diff
@@ -25,12 +25,8 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 int ModApiStorage::l_get_mod_storage(lua_State *L)
 {
-	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_CURRENT_MOD_NAME);
-	if (!lua_isstring(L, -1)) {
-		return 0;
-	}
-
-	std::string mod_name = readParam<std::string>(L, -1);
+	// Note that this is wrapped in Lua, see builtin/[...].lua
```
Comment (sfan5) [inline src/script/lua_api/l_storage.cpp:28]: well ... I mean ...
it would make sense to fill in the actual filename here, wouldn't it?

Comment (TurkeyMcMac) [reply src/script/lua_api/l_storage.cpp:28]: Oops, didn't see that. Haha.

**Review by sfan5 — APPROVED** [review]
LGTM

**Review by SmallJoker — APPROVED** [review]
Works as intended. Code looks good.

**Comment by appgurueu** [comment]
<blockquote>
```lua
assert(core.get_mod_storage() == core.get_mod_storage())                        
collectgarbage()                                                                
assert(core.get_mod_storage())
```
</blockquote>

Perhaps add this to `games/devtest/mods/unittests/misc.lua`?

**Comment by TurkeyMcMac** [comment]
@appgurueu I have now added a unit test as part of https://github.com/minetest/minetest/pull/12562.

**Pull request merged**